File: java/eshop-cqrs-ddd-microservices-excercise/common/src/main/java/com/marekturis/common/infrastructure/messaging/RabbitMQEventPublisher.java
package com.marekturis.common.infrastructure.messaging;

import com.marekturis.common.application.EventJsonSerializer;
import com.marekturis.common.domain.event.Event;
import com.marekturis.common.domain.event.EventHandler;
import com.marekturis.common.domain.event.EventPublisher;
import com.rabbitmq.client.*;

import javax.inject.Inject;
import javax.inject.Named;
import java.io.IOException;
import java.util.concurrent.TimeoutException;

/**
 * @author Marek Turis
 */
@Named
public class RabbitMQEventPublisher implements EventPublisher, AutoCloseable {

	private static final String EXCHANGE_NAME = "eshop_events";
	private final Connection connection;
	private Channel channel;
	private EventJsonSerializer serializer;

	@Inject
	public RabbitMQEventPublisher(EventJsonSerializer serializer) throws IOException, TimeoutException {
		this.serializer = serializer;

		ConnectionFactory factory = new ConnectionFactory();
		factory.setHost("localhost");
		connection = factory.newConnection();
		channel = connection.createChannel();
		channel.exchangeDeclare(EXCHANGE_NAME, "direct");
	}


	public void registerHandler(EventHandler handler) {
		try {
			String queueName = channel.queueDeclare().getQueue();
			channel.queueBind(queueName, EXCHANGE_NAME, handler.eventToHandle());

			Consumer consumer = new RabbitMQConsumer(channel, handler);
			channel.basicConsume(queueName, consumer);
		} catch (IOException e) {
			throw new RabbitMQFailureException(e);
		}
	}

	public void publish(Event event) {
		String jsonEvent = serializer.fromEvent(event);
		try {
			channel.basicPublish(EXCHANGE_NAME, event.name(), null, jsonEvent.getBytes());
		} catch (IOException e) {
			throw new RabbitMQFailureException(e);
		}
	}

	public void close() throws IOException, TimeoutException {
		if (channel != null && channel.isOpen()) {
			channel.close();
		}

		if (connection != null && connection.isOpen()) {
			connection.close();
		}
	}
}

File: dotnet/eshop-cqrs-ddd-microservices-excercise/Com.Marekturis.Common/infrastructure/RabbitMQEventPublisher.cs
using System;
using Com.Marekturis.Common.Application;
using Com.Marekturis.Common.Domain;
using RabbitMQ.Client;
using EventHandler = Com.Marekturis.Common.Domain.EventHandler;

namespace Com.Marekturis.Common.infrastructure
{
    public class RabbitMQEventPublisher : EventPublisher
    {
        private const string EXCHANGE_NAME = "eshop_events";

        private readonly IConnection connection;
        private readonly IModel channel;

        private EventJsonSerializer serializer;

        public RabbitMQEventPublisher(EventJsonSerializer serializer)
        {
            this.serializer = serializer;

            var factory = new ConnectionFactory {HostName = "localhost"};
            connection = factory.CreateConnection();
            channel = connection.CreateModel();
            channel.ExchangeDeclare(EXCHANGE_NAME, "direct");
        }

        public void RegisterHandler(EventHandler eventHandler)
        {
            var queueName = channel.QueueDeclare().QueueName;

            channel.QueueBind(
                queue: queueName,
                exchange: EXCHANGE_NAME,
                routingKey: eventHandler.EventToHandle);


            IBasicConsumer consumer = new RabbitMQConsumer(eventHandler, channel);
            channel.BasicConsume(queue: queueName, consumer: consumer);
        }

        public void Publish(Event e)
        {
            var jsonEvent = serializer.FromEvent(e);
            channel.BasicPublish(EXCHANGE_NAME, e.Name, null, System.Text.Encoding.UTF8.GetBytes(jsonEvent));
        }

        public void Dispose()
        {
            if (channel != null && channel.IsOpen)
            {
                channel.Close();
            }

            if (connection != null && connection.IsOpen)
            {
                connection.Close();
            }
        }
    }
}